sql/src/main/java/io/crate/analyze/MultiSourceSelect.java

```diff
@@ -102,36 +76,79 @@ public MultiSourceSelect(
 
     @Override
     public Field getField(Path path, Operation operation) throws UnsupportedOperationException {
-        throw new UnsupportedOperationException("getField on SelectAnalyzedStatement is not implemented");
+        if (operation != Operation.READ) {
+            throw new UnsupportedOperationException("getField on MultiSourceSelect is only supported for READ operations");
+        }
+        return fields.get(path.outputName());
     }
 
     @Override
     public List<Field> fields() {
-        return fields;
+        return fields.asList();
     }
 
     public QuerySpec querySpec() {
         return querySpec;
     }
 
-    private RelationSplitter initSources(Map<QualifiedName, AnalyzedRelation> sources) {
+    private HashMap<QualifiedName, Source> initializeSources(Map<QualifiedName, AnalyzedRelation> originalSources) {
+        HashMap<QualifiedName, Source> sources = new LinkedHashMap<>(originalSources.size());
+        for (Map.Entry<QualifiedName, AnalyzedRelation> entry : originalSources.entrySet()) {
+            Source source = new Source(entry.getValue());
+            sources.put(entry.getKey(), source);
+        }
+        return sources;
+    }
 
-        RelationSplitter splitter = RelationSplitter.process(querySpec, sources.values());
-        for (Map.Entry<QualifiedName, AnalyzedRelation> entry : sources.entrySet()) {
-            QuerySpec spec = splitter.getSpec(entry.getValue());
+    public void pushDownQuerySpecs() {
+        splitter.process();
+        for (Source source : sources.values()) {
+            QuerySpec spec = splitter.getSpec(source.relation());
             assert spec != null;
-            Source source = new Source(entry.getValue(), spec);
-            this.sources.put(entry.getKey(), source);
+            source.querySpec(spec);
         }
-        return splitter;
     }
 
     /**
      * Returns an orderBy containing only orderings which are not already pushed down to sources. The optional is only
-     * pressent if there are orderBySymbols left. The existing orderBy is returned if it
+     * present if there are orderBySymbols left. The existing orderBy is returned if it
      * there are no changes.
      */
     public Optional<OrderBy> remainingOrderBy() {
         return splitter.remainingOrderBy();
     }
+
+    public static class Source {
```

Comment:
please avoid moving things in the future if not really necessary. Makes it harder to review